File: examples_stardust/1_advanced/2_nft_owns_did.rs
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::path::PathBuf;

use identity_stardust::NetworkName;
use identity_stardust::StardustClientExt;
use identity_stardust::StardustDocument;

use examples::create_did_document;
use examples::get_address_with_funds;
use examples::NETWORK_ENDPOINT;
use identity_stardust::block::address::NftAddress;
use identity_stardust::block::output::AliasOutput;
use identity_stardust::StardustIdentityClientExt;
use iota_client::api_types::responses::OutputResponse;
use iota_client::block::address::Address;
use iota_client::block::output::unlock_condition::AddressUnlockCondition;
use iota_client::block::output::NftId;
use iota_client::block::output::NftOutput;
use iota_client::block::output::NftOutputBuilder;
use iota_client::block::output::Output;
use iota_client::block::output::OutputId;
use iota_client::block::output::RentStructure;
use iota_client::block::output::UnlockCondition;
use iota_client::block::payload::transaction::TransactionEssence;
use iota_client::block::payload::Payload;
use iota_client::block::Block;
use iota_client::secret::stronghold::StrongholdSecretManager;
use iota_client::secret::SecretManager;
use iota_client::Client;

/// Demonstrates how an identity can be owned by NFTs,
/// and how observers can verify that relationship.
///
/// For this example, we consider the case where a car's NFT owns
/// the DID of the car, so that transferring the NFT also transfers DID ownership.
#[tokio::main]
async fn main() -> anyhow::Result<()> {
  // =============================
  // Create the car's NFT and DID.
  // =============================

  // Create a new client to interact with the IOTA ledger.
  let client: Client = Client::builder().with_primary_node(NETWORK_ENDPOINT, None)?.finish()?;

  // Create a new secret manager backed by a Stronghold.
  let mut secret_manager: SecretManager = SecretManager::Stronghold(
    StrongholdSecretManager::builder()
      .password("secure_password")
      .try_build(PathBuf::from("./example.stronghold"))?,
  );

  // Get an address with funds for testing.
  let address: Address = get_address_with_funds(&client, &mut secret_manager).await?;

  // Get the current byte cost.
  let rent_structure: RentStructure = client.get_rent_structure().await?;

  // Create the car NFT with an Ed25519 address as the unlock condition.
  let car_nft: NftOutput = NftOutputBuilder::new_with_minimum_storage_deposit(rent_structure.clone(), NftId::null())?
    .add_unlock_condition(UnlockCondition::Address(AddressUnlockCondition::new(address)))
    .finish()?;

  // Publish the NFT output.
  let block: Block = client
    .block()
    .with_secret_manager(&secret_manager)
    .with_outputs(vec![car_nft.into()])?
    .finish()
    .await?;
  let _ = client.retry_until_included(&block.id(), None, None).await?;

  let car_nft_id: NftId = NftId::from(get_nft_output_id(
    block
      .payload()
      .ok_or_else(|| anyhow::anyhow!("expected the block to contain a payload"))?,
  )?);

  let network: NetworkName = client.network_name().await?;

  // Construct a DID document for the subsidiary.
  let document: StardustDocument = create_did_document(&network)?;

  // Create a new DID for the car that is owned by the car NFT.
  let car_did_output: AliasOutput = client
    .new_did_output(Address::Nft(car_nft_id.into()), document, Some(rent_structure))
    .await?;

  let car_document: StardustDocument = client.publish_did_output(&secret_manager, car_did_output).await?;

  // ============================================
  // Determine the car's NFT given the car's DID.
  // ============================================

  let output: AliasOutput = client.resolve_did_output(car_document.id()).await?;

  // Extract the NFT address from the state controller unlock condition.
  let unlock_condition: &UnlockCondition = output
    .unlock_conditions()
    .iter()
    .next()
    .ok_or_else(|| anyhow::anyhow!("expected at least one unlock condition"))?;

  let car_nft_address: NftAddress =
    if let UnlockCondition::StateControllerAddress(state_controller_unlock_condition) = unlock_condition {
      if let Address::Nft(nft_address) = state_controller_unlock_condition.address() {
        *nft_address
      } else {
        anyhow::bail!("expected an NFT address as the unlock condition");
      }
    } else {
      anyhow::bail!("expected an Address as the unlock condition");
    };

  // Retrieve the NFT Output of the car.
  let car_nft_id: &NftId = car_nft_address.nft_id();
  let output_id: OutputId = client.nft_output_id(*car_nft_id).await?;
  let output_response: OutputResponse = client.get_output(&output_id).await?;
  let output: Output = Output::try_from(&output_response.output)?;

  let car_nft: NftOutput = if let Output::Nft(nft_output) = output {
    nft_output
  } else {
    anyhow::bail!("expected an NFT output");
  };

  println!("The car's DID is: {car_document:#}");
  println!("The car's NFT is: {car_nft:#?}");

  Ok(())
}

// Helper function to get the output id for the first NFT output in a Block.
fn get_nft_output_id(payload: &Payload) -> anyhow::Result<OutputId> {
  match payload {
    Payload::Transaction(tx_payload) => {
      let TransactionEssence::Regular(regular) = tx_payload.essence();
      for (index, output) in regular.outputs().iter().enumerate() {
        if let Output::Nft(_nft_output) = output {
          return Ok(OutputId::new(tx_payload.id(), index.try_into().unwrap())?);
        }
      }
      anyhow::bail!("no NFT output in transaction essence")
    }
    _ => anyhow::bail!("No transaction payload"),
  }
}

File: bindings/wasm/examples/src/1_advanced/2_nft_owns_did.ts
// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Client, MnemonicSecretManager } from "@iota/client-wasm/node";
import { Bip39 } from "@iota/crypto.js";
import { IotaDocument, IotaIdentityClient } from "@iota/identity-wasm/node";
import {
    ADDRESS_UNLOCK_CONDITION_TYPE,
    AddressTypes,
    Bech32Helper,
    IAliasOutput,
    INftOutput,
    IOutputResponse,
    IRent,
    IStateControllerAddressUnlockCondition,
    ITransactionPayload,
    NFT_ADDRESS_TYPE,
    NFT_OUTPUT_TYPE,
    OutputTypes,
    PayloadTypes,
    STATE_CONTROLLER_ADDRESS_UNLOCK_CONDITION_TYPE,
    TRANSACTION_ESSENCE_TYPE,
    TRANSACTION_PAYLOAD_TYPE,
    TransactionHelper,
} from "@iota/iota.js";
import { Converter } from "@iota/util.js";
import { API_ENDPOINT, ensureAddressHasFunds } from "../util";

/** Demonstrates how an identity can be owned by NFTs,
and how observers can verify that relationship.

For this example, we consider the case where a car's NFT owns
the DID of the car, so that transferring the NFT also transfers DID ownership. */
export async function nftOwnsDid() {
    // =============================
    // Create the car's NFT and DID.
    // =============================

    // Create a new Client to interact with the IOTA ledger.
    const client = new Client({
        primaryNode: API_ENDPOINT,
        localPow: true,
    });
    const didClient = new IotaIdentityClient(client);

    // Generate a random mnemonic for our wallet.
    const secretManager: MnemonicSecretManager = {
        mnemonic: Bip39.randomMnemonic(),
    };

    // Get the current byte costs.
    const rentStructure: IRent = await didClient.getRentStructure();

    // Get the Bech32 human-readable part (HRP) of the network.
    const networkName: string = await didClient.getNetworkHrp();

    // Create a new address that will own the NFT.
    const addressBech32 = (await client.generateAddresses(secretManager, {
        accountIndex: 0,
        range: {
            start: 0,
            end: 1,
        },
    }))[0];
    const address = Bech32Helper.addressFromBech32(addressBech32, networkName);

    // Get funds for testing from the faucet.
    await ensureAddressHasFunds(client, addressBech32);

    // Create the car NFT with an Ed25519 address as the unlock condition.
    let carNft: INftOutput = await client.buildNftOutput({
        nftId: "0x0000000000000000000000000000000000000000000000000000000000000000",
        unlockConditions: [
            {
                // The NFT will initially be owned by the Ed25519 address.
                type: ADDRESS_UNLOCK_CONDITION_TYPE,
                address,
            },
        ],
    });

    // Set the appropriate storage deposit.
    carNft.amount = TransactionHelper.getStorageDeposit(carNft, rentStructure).toString();

    // Publish the NFT.
    const [blockId, block] = await client.buildAndPostBlock(secretManager, { outputs: [carNft] });
    await client.retryUntilIncluded(blockId);

    // Extract the identifier of the NFT from the published block.
    // Non-null assertion is safe because we published a block with a payload.
    var carNftId: string = nft_output_id(block.payload!);

    // Create the address of the NFT.
    const nftAddress: AddressTypes = {
        type: NFT_ADDRESS_TYPE,
        nftId: carNftId,
    };

    // Construct a DID document for the car.
    var carDocument: IotaDocument = new IotaDocument(networkName);

    // Create a new DID for the car that is owned by the car NFT.
    var carDidAliasOutput: IAliasOutput = await didClient.newDidOutput(nftAddress, carDocument, rentStructure);

    // Publish the car DID.
    carDocument = await didClient.publishDidOutput(secretManager, carDidAliasOutput);

    // ============================================
    // Determine the car's NFT given the car's DID.
    // ============================================

    // Resolve the Alias Output of the DID.
    carDidAliasOutput = await didClient.resolveDidOutput(carDocument.id());

    // Extract the NFT Id from the state controller unlock condition.
    const stateControllerUnlockCondition: IStateControllerAddressUnlockCondition = carDidAliasOutput.unlockConditions
        .find(feature =>
            feature.type === STATE_CONTROLLER_ADDRESS_UNLOCK_CONDITION_TYPE
        ) as IStateControllerAddressUnlockCondition;
    if (stateControllerUnlockCondition.address.type === NFT_ADDRESS_TYPE) {
        carNftId = stateControllerUnlockCondition.address.nftId;
    } else {
        throw new Error("expected nft address unlock condition");
    }

    // Fetch the NFT Output of the car.
    const nftOutputId: string = await client.nftOutputId(carNftId);
    const outputResponse: IOutputResponse = await client.getOutput(nftOutputId);
    const output: OutputTypes = outputResponse.output;

    if (output.type === NFT_OUTPUT_TYPE) {
        carNft = output;
    } else {
        throw new Error("expected nft output type");
    }

    console.log("The car's DID is:", JSON.stringify(carDocument, null, 2));
    console.log("The car's NFT is:", JSON.stringify(carNft, null, 2));
}

function nft_output_id(payload: PayloadTypes): string {
    if (payload.type === TRANSACTION_PAYLOAD_TYPE) {
        const txPayload: ITransactionPayload = payload;
        const txHash = Converter.bytesToHex(TransactionHelper.getTransactionPayloadHash(txPayload), true);

        if (txPayload.essence.type === TRANSACTION_ESSENCE_TYPE) {
            const outputs = txPayload.essence.outputs;
            for (let index in txPayload.essence.outputs) {
                if (outputs[index].type === NFT_OUTPUT_TYPE) {
                    const outputId: string = TransactionHelper.outputIdFromTransactionData(txHash, parseInt(index));
                    return TransactionHelper.resolveIdFromOutputId(outputId);
                }
            }
            throw new Error("no NFT output in transaction essence");
        } else {
            throw new Error("expected transaction essence");
        }
    } else {
        throw new Error("expected transaction payload");
    }
}
